metricbeat/module/kubernetes/util/kubernetes.go

```diff
@@ -64,18 +52,43 @@ type kubernetesConfig struct {
 	Namespace           string                              `config:"namespace"`
 }
 
+// Enricher takes Kubernetes events and enrich them with k8s metadata
+type Enricher interface {
+	// Start will start the Kubernetes watcher on the first call, does nothing on the rest
+	// errors are logged as warning
+	Start(*Watchers)
+
+	// Stop will stop the Kubernetes watcher
+	Stop(*Watchers)
+
+	// Enrich the given list of events
+	Enrich([]mapstr.M)
+}
+
 type enricher struct {
 	sync.RWMutex
-	metadata            map[string]mapstr.M
-	index               func(mapstr.M) string
-	watcher             kubernetes.Watcher
-	watchersStarted     bool
-	watchersStartedLock sync.Mutex
-	namespaceWatcher    kubernetes.Watcher
-	nodeWatcher         kubernetes.Watcher
-	replicasetWatcher   kubernetes.Watcher
-	jobWatcher          kubernetes.Watcher
-	isPod               bool
+	metadata     map[string]mapstr.M
+	index        func(mapstr.M) string
+	resourceName string
+	isPod        bool
+	config       *kubernetesConfig
+}
+
+type nilEnricher struct{}
+
+func (*nilEnricher) Start(*Watchers)   {}
+func (*nilEnricher) Stop(*Watchers)    {}
+func (*nilEnricher) Enrich([]mapstr.M) {}
+
+type watcherData struct {
+	whichAreUsing []string // list of resources using this watcher
```

Comment:
I believe that `resourcesUsing` could be a better name of that array and for all the associated functions of additions and removals from that array(e.g. addToWhichAreUsing --> addResourcesUsing)